docs/RFCS/sql_column_groups.md

```diff
@@ -0,0 +1,122 @@
+to the table.
+
+Note that there is also per key overhead at the transaction
+level. Every key written within a transaction has a "write intent"
+associated with it and these intents need to be resolved when the
+transaction is committed. Reducing the number of keys per row reduces
+this overhead.
+
+# Detailed design
+
+The structure for table keys will be changed to:
+
+```
+/<tableID>/<indexID>/<primaryKeyColumns...>/<groupID>
+```
+
+The value proto will remain almost the same, but the `tag` field will
+be removed and instead of encoding the column value directly into the
+`raw_bytes` field we'll encoding a series of `<columnID>/<columnVal>`
+pairs where `<columnID>` is varint encoded and `<columnVal>` is
+encoded per its type using the `util/encoding` routines. Similar to
+the existing convention, `NULL` column values will not be present in
+the value.
+
+The `ColumnDescriptor` proto will be extended with a `GroupID`
+field. `GroupIDs` will be allocated per-table using a new
+`TableDescriptor.NextGroupID` field. `GroupID(0)` will be always be
+written for a row and will take the place of the sentinel key. We
+could even omit the `/<groupID>` key suffix when encoding `GroupID(0)`
+to make the encoding identical to the previous sentinel key encoding.
+
+When a column is added to a table it will be assigned to a group. A
+group will be selected so that raw column data size for the columns in
+the group is less than some fixed size (e.g. 256 bytes). Note that
+assigning every column to its own group would be effectively
+equivalent to the current SQL to KV mapping.
+
+# Drawbacks
+
+* Grouping multiple columns into a single key/value pair will
+necessitate changes to scan, insert, update and delete. Update will
+require special attention as the `NULL`-ing of a column will be
+performed via a `Put` instead of a `Del`.
+
+# Alternatives
+
+# Unresolved questions
+
+# Performance Experiments
+
+For the above `kv` table, we can approximate the benefit of this
+change for a benchmark by not writing the sentinel key for each
+row. The numbers below show the delta for that change using the `kv`
+table structure described above (instead of the 1 column table
+currently used in the `{Insert,Scan}` benchmarks).
+
+```
+name                   old time/op    new time/op    delta
+Insert1_Cockroach-8       983µs ± 1%     948µs ± 0%   -3.53%    (p=0.000 n=9+9)
+Insert10_Cockroach-8     1.72ms ± 1%    1.34ms ± 0%  -22.05%   (p=0.000 n=10+9)
+Insert100_Cockroach-8    8.52ms ± 1%    4.99ms ± 1%  -41.42%  (p=0.000 n=10+10)
+Scan1_Cockroach-8         348µs ± 1%     345µs ± 1%   -1.07%  (p=0.002 n=10+10)
+Scan10_Cockroach-8        464µs ± 1%     419µs ± 1%   -9.68%  (p=0.000 n=10+10)
+Scan100_Cockroach-8      1.33ms ± 1%    0.95ms ± 1%  -28.61%  (p=0.000 n=10+10)
+```
+
+While the benefit is fairly small for single row insertions, this is
+only benchmarking the simplest of tables. We'd expect a bigger benefit
+for tables with more columns.
```

Comment:
I'm not sure about this. The sentinel row is nearly half the (pre-prefix-compression) size of this kv table, but it is a smaller proportion of the data stored for a larger table. This is also a best-case scenario in that there are no variable-size columns. Insertions and `SELECT *` are faster, but updates have to read and rewrite unrelated columns. I think it's still a performance win, but I'm not sure if it's a big enough win to justify the complexity of such an intrusive change.